sdks/java/core/src/main/java/org/apache/beam/sdk/io/ReadAllViaFileBasedSource.java

```diff
@@ -98,15 +108,53 @@ private ReadFileRangesFn(
     public void process(ProcessContext c) throws IOException {
       ReadableFile file = c.element().getKey();
       OffsetRange range = c.element().getValue();
-      FileBasedSource<T> source =
+      c.output(
           CompressedSource.from(createSource.apply(file.getMetadata().resourceId().toString()))
-              .withCompression(file.getCompression());
+              .withCompression(file.getCompression())
+              .createForSubrangeOfFile(file.getMetadata(), range.getFrom(), range.getTo()));
+    }
+  }
+
+  /**
+   * A {@link Coder} for {@link BoundedSource}s that wraps a {@link SerializableCoder}. We cannot
+   * safely use an unwrapped SerializableCoder because {@link
+   * SerializableCoder#structuralValue(Serializable)} assumes that coded elements support object
+   * equality (https://issues.apache.org/jira/browse/BEAM-3807). By default, Coders compare equality
+   * by serialized bytes, which we want in this case. It is usually safe to depend on coded
+   * representation here because we only compare objects on bundle commit, which compares
+   * serializations of the same object instance.
+   *
+   * <p>BoundedSources are generally not used as PCollection elements, so we do not expose this
+   * coder for wider use.
+   */
+  public static class BoundedSourceCoder<T> extends CustomCoder<BoundedSource<T>> {
```

Comment:
Wondering if I should put thisas an inner class of BoundedSource or other place or if is ok here..
Also if we should auto register this Coder in `BoundedSource`.

Reply:
Probably this should be it's own file if public ?

I don't think there's a reason to register since I don't think using BoundedSources as elements of a PCollection should be a recommended pattern for Beam (even though we do it here as a stop gap till we get SDF).

Reply:
Yes that's exactly what I thought. I mean I over refactored in this PR I can just let things as they were, where this is done just at the `runners/core-construction-java` level so no need to do that.